### PR TITLE
Add `Withdrawals` struct

### PR DESCRIPTION
**Pull request by tcoratger**
## Changes Proposed

- Added a `Withdrawals` structure in the `primitives` module to encapsulate a vector of `Withdrawal`.
- Noticed that an equivalent type, `Withdrawal`, is also present in the `alloy` library but resides in the `rpc` crate. This often leads to confusion between the `Withdrawal` type of reth primitives and the `Withdrawal` type of alloy.
- To achieve complete integration and unify the types, suggesting the following steps:
    - Move `Withdrawal` and `Withdrawals` up a level in the `alloy` crate to place them at the root instead of within the `rpc` module.
    - Completely remove the definitions of `Withdrawal` and `Withdrawals` from the reth primitives.

## Next Steps

If this pull request is accepted, I propose to merge it first. 

Following that, further work could be carried out on the `alloy` library to complete the refactoring. 

The aim is to rely solely on `Withdrawal` and `Withdrawals` from the `alloy` library, eliminating the need for these types in the reth primitives.

### CODE REVIEW

**Comment by tcoratger** [comment]
@mattsse wdt about this one? If this is fine I can work on rlp decoding stuff for `Block` in order to pass the CI

File: crates/storage/provider/src/test_utils/blocks.rs [hunk 1]
```diff
@@ -81,15 +81,15 @@ pub fn genesis() -> SealedBlock {
...
     let mut block = SealedBlock::decode(&mut block_rlp).unwrap();
-    block.withdrawals = Some(vec![Withdrawal::default()]);
+    block.withdrawals = Some(Withdrawals::default());
```
Comment (mattsse) [inline crates/storage/provider/src/test_utils/blocks.rs:92]: this doesn't look equivalent

File: crates/storage/provider/src/test_utils/blocks.rs [hunk 2]
```diff
@@ -155,7 +155,7 @@ fn block2(
...
```
Comment (mattsse) [inline crates/storage/provider/src/test_utils/blocks.rs:158]: same

File: crates/interfaces/src/p2p/full_block.rs [hunk 3]
```diff
@@ -320,7 +320,7 @@ fn ensure_valid_body_response(
         ))
     }
 
-    let withdrawals = block.withdrawals.as_deref().unwrap_or(&[]);
+    let withdrawals = &block.withdrawals.clone().unwrap_or_default();
```
Comment (mattsse) [inline crates/interfaces/src/p2p/full_block.rs:323]: I think we can do without the clone by adding as_slice

**Review by mattsse — CHANGES_REQUESTED** [review]
I'm okay with this and like this more.

wdyt @rkrasiuk @Rjected

File: crates/storage/provider/src/test_utils/blocks.rs (hunk 2 above)
Comment (tcoratger) [reply crates/storage/provider/src/test_utils/blocks.rs]: Fixed

File: crates/primitives/src/withdrawal.rs [hunk 4]
```diff
@@ -33,6 +37,73 @@ impl Withdrawal {
...
+    fn length(&self) -> usize {
+        self.0.length()
+    }
+}
+
+impl Decodable for Withdrawals {
+    fn decode(buf: &mut &[u8]) -> alloy_rlp::Result<Self> {
+        Ok(Self(Decodable::decode(buf)?))
+    }
+}
```
Comment (Rjected) [inline crates/primitives/src/withdrawal.rs:105]: This can be done with `RlpEncodableWrapper` and `RlpDecodableWrapper`

**Review by Rjected — CHANGES_REQUESTED** [review]
Having a `Withdrawals` type is reasonable, just one comment about rlp derive options

File: crates/primitives/src/withdrawal.rs (hunk 4 above)
Comment (tcoratger) [reply crates/primitives/src/withdrawal.rs]: Nice thanks, just fixed, didn't know that

**Comment by tcoratger** [comment]
@Rjected anything to unlock here?

**Review by Rjected — APPROVED** [review]
This looks good to me now

**Review by mattsse — APPROVED** [review]

**Pull request merged**